### ISSUE TITLE
Switch case doesn't match anymore in VS2019

### ISSUE THREAD

**Issue by veloek** [dotnet/roslyn#35661]
**Version Used**: 
Visual Studio 2019
**Example code**
```
public Type GetIntArrayType(object obj)
{
    switch (obj)
    {
        case int[] _ when obj.GetType() != typeof(uint[]):
            return typeof(int[]);
        case uint[] _:
            return typeof(uint[]);
        default:
            throw new Exception("Not an int array");
    }
}

Console.WriteLine(GetIntArrayType(new uint[0]));
```
**Expected Behavior**:
`System.UInt32[]`
**Actual Behavior**:
`System.Exception : Not an int array`

This code worked in VS2017, but no longer in VS2019. I suspect some optimization was done to skip cases that "can't match", but in this rather special case it is mistaken.

It will match `int[]`(!) but fail the condition and instead of testing the next case it jumps to default.

The fact that an uint array can be unboxed as an int array is strange as well, but the reason behind that is explained here: https://blogs.msdn.microsoft.com/ericlippert/2009/09/24/why-is-covariance-of-value-typed-arrays-inconsistent/

I understand that this is an edge case that you might not deem important enough to do anything about, but I wanted to let you know about this change in behavior from VS2017.

**Comment by jcouv** [comment]
Tagging @gafter

**Comment by gafter** [comment]
This is fixed in #35772.

**Comment by veloek** [comment]
Thanks for jumping on this and fixing it so quickly 🙂 Keep up the great work!
